File: python/tests/rubycore/test_cross_language_validation.py
"""Cross-language validation tests for Marshal implementations.

This module provides tests that validate the Python Marshal implementation produces
identical results to the Ruby reference implementation, ensuring 100% compatibility.
"""

import json
import subprocess
import pytest
from pathlib import Path
from typing import Any, Dict

from railscompat.rubycore.marshal import Marshal


class TestCrossLanguageValidation:
    """Cross-language validation tests for Marshal compatibility."""

    @pytest.fixture(scope='class')
    def validator_script(self) -> Path:
        """Path to Ruby cross-language validator script."""
        script_path = Path(__file__).parent.parent.parent.parent / 'shared' / 'ruby-reference' / 'cross_language_validator.rb'
        
        if not script_path.exists():
            pytest.skip(f"Validator script not found: {script_path}")
        
        return script_path

    @pytest.fixture(scope='class')
    def test_cases(self) -> Dict[str, Any]:
        """Load test cases for cross-language validation."""
        test_data_path = Path(__file__).parent.parent.parent.parent / 'shared' / 'test-data' / 'marshal_test_cases.json'
        
        if not test_data_path.exists():
            pytest.skip(f"Test data file not found: {test_data_path}")
        
        with open(test_data_path, 'r') as f:
            data = json.load(f)
        
        return data['test_cases']

    def _run_ruby_validator(self, validator_script: Path, command: str, *args) -> Dict[str, Any]:
        """Run Ruby validator script and return parsed JSON result."""
        cmd = ['ruby', str(validator_script), command] + list(args)
        
        try:
            result = subprocess.run(cmd, capture_output=True, text=True, check=True)
            return json.loads(result.stdout)
        except subprocess.CalledProcessError as e:
            pytest.fail(f"Ruby validator failed: {e.stderr}")
        except json.JSONDecodeError as e:
            pytest.fail(f"Failed to parse validator output: {e}")

    def test_ruby_validator_available(self, validator_script: Path) -> None:
        """Test that Ruby validator script is available and working."""
        result = self._run_ruby_validator(validator_script, 'validate', 'nil_value')
        
        assert 'case_name' in result
        assert result['case_name'] == 'nil_value'
        assert 'matches' in result

    @pytest.mark.parametrize("case_name", [
        "nil_value", "true_value", "false_value", "zero", "small_positive",
        "simple_string", "simple_symbol", "empty_array", "simple_array",
        "empty_hash", "simple_hash"
    ])
    def test_basic_cross_language_compatibility(
        self, 
        validator_script: Path, 
        test_cases: Dict[str, Any], 
        case_name: str
    ) -> None:
        """Test that Python implementation matches Ruby for basic types."""
        # Get test case data
        test_case = test_cases[case_name]
        marshal_base64 = test_case['marshal_base64']
        
        # Parse with Python implementation
        python_result = Marshal.load_b64(marshal_base64)
        
        # Validate with Ruby reference
        ruby_validation = self._run_ruby_validator(validator_script, 'validate', case_name)
        
        # Ruby validator should confirm the data is valid
        if 'error' in ruby_validation:
            pytest.fail(f"Ruby validation failed for {case_name}: {ruby_validation['error']}")
        
        assert ruby_validation['matches'] is True, f"Ruby validation failed for {case_name}"
        
        # Compare Python result with expected result
        expected = test_case['expected_result']
        
        # Handle symbol conversion (Ruby symbols become strings with ':' prefix)
        if isinstance(expected, str) and case_name.startswith('symbol_'):
            # For symbols, Python should return string with ':' prefix
            assert python_result.startswith(':'), f"Symbol should start with ':' for {case_name}"
        else:
            # For other types, direct comparison
            assert python_result == expected, f"Python result doesn't match expected for {case_name}"

    def test_complex_structure_compatibility(
        self, 
        validator_script: Path, 
        test_cases: Dict[str, Any]
    ) -> None:
        """Test complex nested structures for cross-language compatibility."""
        complex_cases = ['nested_hash', 'mixed_array', 'rails_session_simple']
        
        for case_name in complex_cases:
            if case_name not in test_cases:
                continue
                
            test_case = test_cases[case_name]
            marshal_base64 = test_case['marshal_base64']
            
            # Parse with Python implementation
            python_result = Marshal.load_b64(marshal_base64)
            
            # Validate with Ruby reference
            ruby_validation = self._run_ruby_validator(validator_script, 'validate', case_name)
            
            # Ruby validator should confirm the data is valid
            if 'error' in ruby_validation:
                pytest.fail(f"Ruby validation failed for {case_name}: {ruby_validation['error']}")
            
            assert ruby_validation['matches'] is True, f"Ruby validation failed for {case_name}"
            
            # Basic structure validation for Python result
            assert python_result is not None, f"Python result is None for {case_name}"

    def test_symbol_link_compatibility(
        self, 
        validator_script: Path, 
        test_cases: Dict[str, Any]
    ) -> None:
        """Test symbol link (reused symbols) compatibility."""
        if 'symbol_reuse' not in test_cases:
            pytest.skip("symbol_reuse test case not available")
        
        test_case = test_cases['symbol_reuse']
        marshal_base64 = test_case['marshal_base64']
        
        # Parse with Python implementation
        python_result = Marshal.load_b64(marshal_base64)
        
        # Validate with Ruby reference
        ruby_validation = self._run_ruby_validator(validator_script, 'validate', 'symbol_reuse')
        
        if 'error' in ruby_validation:
            pytest.fail(f"Ruby validation failed for symbol_reuse: {ruby_validation['error']}")
        
        assert ruby_validation['matches'] is True, "Ruby validation failed for symbol_reuse"
        
        # Python-specific validation
        assert isinstance(python_result, list), "symbol_reuse should be an array"
        assert len(python_result) == 4, "symbol_reuse should have 4 elements"
        
        # Check symbol reuse (first, second, and fourth elements should be identical)
        assert python_result[0] == python_result[1] == python_result[3], "Reused symbols should be identical"
        assert python_result[2] != python_result[0], "Different symbol should be different"

    def test_generate_custom_case_compatibility(self, validator_script: Path) -> None:
        """Test compatibility with custom generated Ruby cases."""
        # Generate a custom test case using Ruby
        custom_literal = '{"custom" => [1, 2, :symbol], "nested" => {"key" => "value"}}'
        
        ruby_generated = self._run_ruby_validator(validator_script, 'generate', custom_literal)
        
        if 'error' in ruby_generated:
            pytest.skip(f"Ruby generation failed: {ruby_generated['error']}")
        
        # Parse with Python implementation
        python_result = Marshal.load_b64(ruby_generated['marshal_base64'])
        
        # Basic validation - should be a hash with expected keys
        assert isinstance(python_result, dict), "Generated case should be a hash"
        assert 'custom' in python_result, "Should contain 'custom' key"
        assert 'nested' in python_result, "Should contain 'nested' key"

    def test_error_handling_compatibility(self, validator_script: Path) -> None:
        """Test that error handling is consistent between implementations."""
        # Test invalid Base64
        with pytest.raises(Exception):
            Marshal.load_b64("invalid_base64!")
        
        # Test malformed Marshal data
        with pytest.raises(Exception):
            Marshal.load_b64("YWJjZGVmZ2g=")  # "abcdefgh" in base64
        
        # Test empty data
        with pytest.raises(Exception):
            Marshal.load(b"")

    def test_performance_consistency(self, test_cases: Dict[str, Any]) -> None:
        """Test that Python implementation handles same data sizes as Ruby."""
        # Find the largest test case
        largest_case = max(test_cases.items(), key=lambda x: x[1]['data_size'])
        case_name, test_case = largest_case
        
        # Should be able to parse the largest case without issues
        result = Marshal.load_b64(test_case['marshal_base64'])
        assert result is not None, f"Failed to parse largest case: {case_name}"

    def test_all_cases_cross_validation(
        self, 
        validator_script: Path, 
        test_cases: Dict[str, Any]
    ) -> None:
        """Run cross-validation on all available test cases."""
        # Run Ruby validator on all cases
        all_validation = self._run_ruby_validator(validator_script, 'validate')
        
        assert 'metadata' in all_validation, "Should have metadata"
        assert 'results' in all_validation, "Should have results"
        
        metadata = all_validation['metadata']
        results = all_validation['results']
        
        # All cases should pass Ruby validation
        total_cases = metadata['total_cases']
        passed_cases = metadata['passed']
        failed_cases = metadata['failed']
        
        # Report any failures
        if failed_cases > 0:
            failed_case_names = [name for name, result in results.items() 
                                if result.get('matches') is False or 'error' in result]
            pytest.fail(f"Ruby validation failed for {failed_cases}/{total_cases} cases: {failed_case_names}")
        
        assert passed_cases == total_cases, f"Expected all cases to pass, got {passed_cases}/{total_cases}"
        
        # Now test that Python can parse all the same cases
        python_failures = []
        for case_name, test_case_data in test_cases.items():
            try:
                python_result = Marshal.load_b64(test_case_data['marshal_base64'])
                assert python_result is not None
            except Exception as e:
                python_failures.append(f"{case_name}: {e}")
        
        if python_failures:
            pytest.fail(f"Python parsing failed for cases: {python_failures}")

File: java/src/test/java/io/gcxrun/railscompat/rubycore/TestCrossLanguageValidation.java
package io.gcxrun.railscompat.rubycore;

import static org.junit.jupiter.api.Assertions.*;

import java.io.BufferedReader;
import java.io.File;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.json.JSONObject;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.ValueSource;

/**
 * Cross-language validation tests for Marshal compatibility.
 * 
 * These tests validate that the Java Marshal implementation produces identical results
 * to the Ruby reference implementation, ensuring 100% compatibility.
 */
public class TestCrossLanguageValidation {

  /**
   * Run Ruby cross-language validator script and return parsed JSON result.
   */
  private JSONObject runRubyValidator(String command, String... args) throws Exception {
    // Path to Ruby validator script
    File validatorScript = new File("../shared/ruby-reference/cross_language_validator.rb");
    
    if (!validatorScript.exists()) {
      // Try alternative path structure
      validatorScript = new File("../../shared/ruby-reference/cross_language_validator.rb");
    }
    
    if (!validatorScript.exists()) {
      throw new RuntimeException("Ruby validator script not found. Please ensure shared infrastructure is set up.");
    }

    List<String> cmd = new ArrayList<>();
    cmd.add("ruby");
    cmd.add(validatorScript.getAbsolutePath());
    cmd.add(command);
    cmd.addAll(Arrays.asList(args));

    ProcessBuilder pb = new ProcessBuilder(cmd);
    Process process = pb.start();

    // Read stdout
    BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
    StringBuilder output = new StringBuilder();
    String line;
    while ((line = reader.readLine()) != null) {
      output.append(line).append("\n");
    }

    // Read stderr for errors
    BufferedReader errorReader = new BufferedReader(new InputStreamReader(process.getErrorStream()));
    StringBuilder errorOutput = new StringBuilder();
    while ((line = errorReader.readLine()) != null) {
      errorOutput.append(line).append("\n");
    }

    int exitCode = process.waitFor();
    
    if (exitCode != 0) {
      throw new RuntimeException("Ruby validator failed with exit code " + exitCode + ": " + errorOutput.toString());
    }

    return new JSONObject(output.toString().trim());
  }

  @Test
  public void testRubyValidatorAvailable() throws Exception {
    JSONObject result = runRubyValidator("validate", "nil_value");
    
    assertTrue(result.has("case_name"), "Should have case_name field");
    assertEquals("nil_value", result.getString("case_name"));
    assertTrue(result.has("matches"), "Should have matches field");
  }

  @ParameterizedTest
  @ValueSource(strings = {
      "nil_value", "true_value", "false_value", "zero", "small_positive",
      "simple_string", "simple_symbol", "empty_array", "simple_array",
      "empty_hash", "simple_hash"
  })
  public void testBasicCrossLanguageCompatibility(String caseName) throws Exception {
    // Validate with Ruby reference
    JSONObject rubyValidation = runRubyValidator("validate", caseName);
    
    // Ruby validator should confirm the data is valid
    if (rubyValidation.has("error")) {
      fail("Ruby validation failed for " + caseName + ": " + rubyValidation.getString("error"));
    }
    
    assertTrue(rubyValidation.getBoolean("matches"), "Ruby validation failed for " + caseName);
    
    // Parse with Java implementation
    String marshalBase64 = rubyValidation.getString("marshal_base64");
    Object javaResult = Marshal.load(marshalBase64);
    
    // Basic validation - should not be null and should match expected type
    assertNotNull(javaResult, "Java result should not be null for " + caseName);
    
    // Type-specific validations
    Object expectedResult = rubyValidation.get("expected");
    validateResultType(caseName, javaResult, expectedResult);
  }

  private void validateResultType(String caseName, Object javaResult, Object expectedResult) {
    if (expectedResult == null || expectedResult == JSONObject.NULL) {
      assertNull(javaResult, "Expected null for " + caseName);
    } else if (expectedResult instanceof Boolean) {
      assertEquals(expectedResult, javaResult, "Boolean mismatch for " + caseName);
    } else if (expectedResult instanceof Number) {
      assertTrue(javaResult instanceof Long, "Expected Long for integer case " + caseName);
    } else if (expectedResult instanceof String) {
      assertTrue(javaResult instanceof String, "Expected String for " + caseName);
      String expectedStr = (String) expectedResult;
      String actualStr = (String) javaResult;
      
      // Handle symbol conversion (symbols should start with ':')
      if (caseName.contains("symbol")) {
        assertTrue(actualStr.startsWith(":"), "Symbol should start with ':' for " + caseName);
      } else {
        assertEquals(expectedStr, actualStr, "String mismatch for " + caseName);
      }
    } else if (expectedResult instanceof org.json.JSONArray) {
      assertTrue(javaResult instanceof List, "Expected List for array case " + caseName);
    } else if (expectedResult instanceof JSONObject) {
      assertTrue(javaResult instanceof java.util.Map, "Expected Map for hash case " + caseName);
    }
  }

  @Test
  public void testSymbolReuseCompatibility() throws Exception {
    JSONObject rubyValidation = runRubyValidator("validate", "symbol_reuse");
    
    if (rubyValidation.has("error")) {
      fail("Ruby validation failed for symbol_reuse: " + rubyValidation.getString("error"));
    }
    
    assertTrue(rubyValidation.getBoolean("matches"), "Ruby validation failed for symbol_reuse");
    
    // Parse with Java implementation
    String marshalBase64 = rubyValidation.getString("marshal_base64");
    Object javaResult = Marshal.load(marshalBase64);
    
    // Should be a list with 4 elements
    assertTrue(javaResult instanceof List, "symbol_reuse should be List");
    List<?> resultList = (List<?>) javaResult;
    assertEquals(4, resultList.size(), "symbol_reuse should have 4 elements");
    
    // Check symbol reuse (first, second, and fourth elements should be identical)
    assertEquals(resultList.get(0), resultList.get(1), "Reused symbols should be identical");
    assertEquals(resultList.get(0), resultList.get(3), "Reused symbols should be identical");
    assertNotEquals(resultList.get(2), resultList.get(0), "Different symbol should be different");
  }

  @Test
  public void testGenerateCustomCaseCompatibility() throws Exception {
    // Generate a custom test case using Ruby
    String customLiteral = "{\"custom\" => [1, 2, :symbol], \"nested\" => {\"key\" => \"value\"}}";
    
    JSONObject rubyGenerated = runRubyValidator("generate", customLiteral);
    
    if (rubyGenerated.has("error")) {
      // Skip if Ruby generation failed
      System.out.println("Skipping custom case test: " + rubyGenerated.getString("error"));
      return;
    }
    
    // Parse with Java implementation
    String marshalBase64 = rubyGenerated.getString("marshal_base64");
    Object javaResult = Marshal.load(marshalBase64);
    
    // Basic validation - should be a hash with expected keys
    assertTrue(javaResult instanceof java.util.Map, "Generated case should be a Map");
    java.util.Map<?, ?> resultMap = (java.util.Map<?, ?>) javaResult;
    assertTrue(resultMap.containsKey("custom"), "Should contain 'custom' key");
    assertTrue(resultMap.containsKey("nested"), "Should contain 'nested' key");
  }

  @Test
  public void testErrorHandlingConsistency() {
    // Test that error handling is consistent between implementations
    
    // Invalid Base64
    assertThrows(Marshal.MarshalException.class, () -> {
      Marshal.load("invalid_base64!");
    }, "Should throw MarshalException for invalid Base64");
    
    // Malformed Marshal data
    assertThrows(Marshal.MarshalException.class, () -> {
      Marshal.load("YWJjZGVmZ2g="); // "abcdefgh" in base64
    }, "Should throw MarshalException for malformed Marshal data");
    
    // Empty data
    assertThrows(Marshal.MarshalException.class, () -> {
      Marshal.load(new byte[0]);
    }, "Should throw MarshalException for empty data");
  }

  @Test
  public void testAllRubyCasesValidateSuccessfully() throws Exception {
    // Run Ruby validator on all cases
    JSONObject allValidation = runRubyValidator("validate");
    
    assertTrue(allValidation.has("metadata"), "Should have metadata");
    assertTrue(allValidation.has("results"), "Should have results");
    
    JSONObject metadata = allValidation.getJSONObject("metadata");
    JSONObject results = allValidation.getJSONObject("results");
    
    // All cases should pass Ruby validation
    int totalCases = metadata.getInt("total_cases");
    int passedCases = metadata.getInt("passed");
    int failedCases = metadata.getInt("failed");
    
    // Report any failures
    if (failedCases > 0) {
      List<String> failedCaseNames = new ArrayList<>();
      var names = results.names();
      if (names != null) {
        for (int i = 0; i < names.length(); i++) {
          String caseName = names.getString(i);
          JSONObject result = results.getJSONObject(caseName);
          if (!result.optBoolean("matches", false) || result.has("error")) {
            failedCaseNames.add(caseName);
          }
        }
      }
      fail("Ruby validation failed for " + failedCases + "/" + totalCases + " cases: " + failedCaseNames);
    }
    
    assertEquals(totalCases, passedCases, "Expected all cases to pass Ruby validation");
    
    System.out.println("All " + totalCases + " Ruby test cases validated successfully");
  }
}